crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -475,6 +563,123 @@ impl WritebackCache {
         )
     }
 
+    fn multi_get_objects_must_exist(&self, object_ids: &[ObjectID]) -> SuiResult<Vec<Object>> {
+        let objects = self.multi_get_objects(object_ids)?;
+        let mut result = Vec::with_capacity(objects.len());
+        for (i, object) in objects.into_iter().enumerate() {
+            if let Some(object) = object {
+                result.push(object);
+            } else {
+                return Err(SuiError::UserInputError {
+                    error: UserInputError::ObjectNotFound {
+                        object_id: object_ids[i],
+                        version: None,
+                    },
+                });
+            }
+        }
+        Ok(result)
+    }
+
+    fn verify_live_object(obj_ref: &ObjectRef, live_object: &Object) -> SuiResult {
+        if obj_ref.1 != live_object.version() {
+            debug!(
+                "object version unavailable for consumption: {:?} (current: {})",
+                obj_ref,
+                live_object.version()
+            );
+            return Err(SuiError::UserInputError {
+                error: UserInputError::ObjectVersionUnavailableForConsumption {
+                    provided_obj_ref: *obj_ref,
+                    current_version: live_object.version(),
+                },
+            });
+        }
+
+        let live_digest = live_object.digest();
+        if obj_ref.2 != live_digest {
+            debug!("object digest mismatch: {:?} vs {:?}", obj_ref, live_digest);
```

Comment:
Should this be a debug_assert + error? And maybe with a TODO to make this an invariant violation

Reply:
No, this is a normal error case that would be hit if someone sends a transaction to sign that has a correct version but incorrect digest.